### PR TITLE
Starting a refactor

### PR DESCRIPTION
**Pull request by dfm**
_Note_: This pull request is not ready to be merged but this is a good place to discuss/code review as I work on this.

Since we need to split plotting, inference, etc. and allow for stopping and starting, I'm starting this branch to work on that stuff.

### CODE REVIEW

File: code/plotting.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    hw1 = (thisPsf.shape[0] - 1) / 2
+    hw2 = (data.shape[0] - thisPsf.shape[0] - 1) / 2
+    hogg_plot_image(ax1, data, stretch)
+    ax1.set_title("data")
+
+    # NOTE: Hogg, you had `scene` instead of `thisScene` in the line below
+    # which was resolving to the _global_ scene object. I think that
+    # `thisScene` is what we actually want though...
+    hogg_plot_image(ax2, convolve(thisScene, thisPsf, mode="valid"), stretch)
```
Comment (dfm) [inline code/plotting.py]: @davidwhogg I changed my mind... see my comment _above_... :-)

**Comment by drphilmarshall** [comment]
HUGE REFACTOR

From a dumb user point of view, I think I'd like LuckyImage script that took one argument, the name of a directory that contains fits images (and other things), and works on them all (not throwing any out). I think this means we need a way of parsing files for data or var images. I vote for a filename convention, _sci.fits and _var.fits but I suppose reading headers could make more sense.  I guess optional arguments would include centroid and image size, from the looks of what you guys have been doing.

When we start doing multi-filter stuff, prior control could become very important - as may the reconstruction of multiple scenes whose combination (in some way) constitutes the final model image. This is total vaporware, just wanted to flag the possibility of extending L.I. in this Multi-Band Observations direction.

**Comment by dfm** [comment]
@drphilmarshall I agree with your view of the UI and I'm working on that now. I'm on a mission to remove all `MAGIC` from within the code while still leaving some _magic_...

**Comment by dfm** [comment]
So I think that this is working pretty well now.

The main user interface is now something like `bin/lucky /path/to/your/fits/files` and there are a few other command line arguments available if you type `bin/lucky --help`. All of the heavy lifting is implemented in `lucky/lucky.py` and most of that is the same as before but there are a few interesting changes:
1. The code doesn't plot anymore... it just saves the current state (inferred scene and PSF) to a FITS file.
2. The scene is initialized as a centered Gaussian and the centroiding/trimming of the data is done on the fly as each new image comes in.

Still to do:
1. Fix that damn memory leak! Hogg: it's in the inference code somewhere (if I do nothing in the `infer_psf` and `infer_scene` methods then there is no memory problem). I'm going to try to figure out how to fix it without breaking everything.
2. Review the way that I'm doing the initialization/centroiding. I think that it's pretty awesome and the scene "forgets" my initialization within a few images anyways.
3. Figure out why the PSF inference barfs after the first few images if we run the inference on the triple with size=100 instead of size=64. It has something to do with the centroiding but I can't figure out what goes wrong.
4. Write a restart method.

I think that's it for now.

**Comment by davidwhogg** [comment]
centroiding and initialization are the same problem, as we discussed:  You initialize a "scene" and the centroiding is just what you get when you "correlate" the scene with the data.  I recommend doing this correlation in one-d, twice.

**Comment by dfm** [comment]
That's exactly what I've done in this pull request I was just saying that we should code review to make sure that it is working (but I think that it's ok).

**Comment by dfm** [comment]
Here's where I do the initialization of the scene:

https://github.com/davidwhogg/LuckyImaging/blob/refactor/lucky/lucky.py#L118

Then each time a new image come in, I do the two 1D correlations to find out how to trim the data:

https://github.com/davidwhogg/LuckyImaging/blob/refactor/lucky/lucky.py#L76

**Comment by davidwhogg** [comment]
Okay good.  Can you merge, commit, push, pull on broiler and run it?

**Comment by drphilmarshall** [comment]
Seconded. I have a pile of test PS1 images (and a partially operable
scipy installation) ready to go.

On Sun, May 6, 2012 at 7:07 PM, David W. Hogg
reply@reply.github.com
wrote:

File: lucky/lucky.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+                self.tinykernel, mode="full")
+
+        logging.info("Got PSF {0}. Min: {1}, Median: {2}, Max: {3}"
+                .format(newDeconvolvedPsf.shape, newDeconvolvedPsf.min(),
+                    np.median(newDeconvolvedPsf), newDeconvolvedPsf.max()))
+
+        # Save the new PSF.
+        self.psf = newDeconvolvedPsf
+
+    def _infer_scene(self, data):
```
Comment (dfm) [inline lucky/lucky.py:284]: The memory leak is here somewhere and brutal... I'm trying to track it down. Then I'll merge.

**Pull request merged**